evm_arithmetization/src/cpu/kernel/asm/mpt/linked_list/linked_list.asm

```diff
@@ -0,0 +1,982 @@
+    MSTORE_GENERAL
+    %pop2
+    JUMP
+
+/// Called when an account is deleted: it deletes all slots associated with the account.
+global remove_all_account_slots:
+    // stack: addr, retdest
+    PROVER_INPUT(linked_list::remove_address_slots)
+    // pred_ptr/5, retdest
+    %get_valid_slot_ptr
+    // stack: pred_ptr, addr, retdest
+    // First, check that the previous address is not `addr`
+    DUP1 MLOAD_GENERAL
+    // stack: pred_addr, pred_ptr, addr, retdest
+    DUP3 EQ %jumpi(panic)
+    // stack: pred_ptr, addr, retdest
+    DUP1
+
+// Now, while the next address is `addr`, remove the next slot.
+remove_all_slots_loop:
+    // stack: pred_ptr, pred_ptr, addr, retdest
+    %add_const(4) DUP1 MLOAD_GENERAL
+    // stack: cur_ptr, cur_ptr_ptr, pred_ptr, addr, retdest
+    DUP1 %eq_const(@U256_MAX) %jumpi(remove_all_slots_end)
+    DUP1 %add_const(4) MLOAD_GENERAL 
+    // stack: next_ptr, cur_ptr, cur_ptr_ptr, pred_ptr, addr, retdest
+    SWAP1 DUP1
+    // stack: cur_ptr, cur_ptr, next_ptr, cur_ptr_ptr, pred_ptr, addr, retdest
+    MLOAD_GENERAL
+    DUP6 EQ ISZERO %jumpi(remove_all_slots_pop_and_end)
+    
+    // Remove slot: update the value in cur_ptr_ptr, and set cur_ptr+4 to @U256_MAX.
+    // stack: cur_ptr, next_ptr, cur_ptr_ptr, pred_ptr, addr, retdest
+    SWAP2 SWAP1
+    // stack: next_ptr, cur_ptr_ptr, cur_ptr, pred_ptr, addr, retdest
+    MSTORE_GENERAL
+    // stack: cur_ptr, pred_ptr, addr, retdest
+    %add_const(4) PUSH @U256_MAX
+    MSTORE_GENERAL
+    // stack: pred_ptr, addr, retdest
+    DUP1
+    %jump(remove_all_slots_loop)
+
+remove_all_slots_pop_and_end:
+    POP
+remove_all_slots_end:
+    // stack: next_ptr, cur_ptr_ptr, pred_ptr, addr, retdest
+    %pop4 JUMP
+
+%macro remove_all_account_slots
+    %stack (addr) -> (addr, %%after)
+    %jump(remove_all_account_slots)
+%%after:
+%endmacro
+
+%macro read_accounts_linked_list
+    %stack (addr) -> (addr, 0, %%after)
+    %addr_to_state_key
+    %jump(search_account)
+%%after:
+    // stack: account_ptr
+%endmacro
+
+%macro read_storage_linked_list
+    // stack: slot
+    %slot_to_storage_key
+    %address
+    %addr_to_state_key
+    %stack (addr, key) -> (addr, key, 0, %%after)
+    %jump(search_slot)
+%%after:
+    // stack: slot_ptr
+%endmacro
+
+%macro read_storage_linked_list_w_addr
+    // stack: slot, address
+    %slot_to_storage_key
+    SWAP1
+    %addr_to_state_key
+    %stack (addr, key) -> (addr, key, 0, %%after)
+    %jump(search_slot)
+%%after:
+    // stack: slot_ptr
+%endmacro
+
+%macro first_account
+    // stack: empty
+    PUSH @SEGMENT_ACCOUNTS_LINKED_LIST
+    %next_account
+%endmacro
+
+%macro next_account
+    // stack: node_ptr
+    %add_const(3)
+    MLOAD_GENERAL
+    // stack: next_node_ptr
+%endmacro
+
+%macro first_slot
+    // stack: empty
+    PUSH @SEGMENT_STORAGE_LINKED_LIST
+    %next_slot
+%endmacro
+
+%macro next_slot
+    // stack: node_ptr
+    %add_const(4)
```

Comment:
We should define in `kernel/constants/` the `linked_list` related constants and reuse them, for clarity and future-proofing, as those floating `3` / `4` / `5` can be confusing.